### PR TITLE
Optimize location fixture generation

### PR DESCRIPTION
**Pull request by esoergel**
@proteusvacuum 

I profiled this (as described in https://github.com/dimagi/commcare-hq/pull/16793) against the latest master, and the improvements with that test user are marginal.  I did log the number of queries performed during tests locally, and that's gone down dramatically.  I expect these improvements will be more noticeable for certain specific instances - users with multiple assigned locations, include_without_expanding set to a relatively numerous level, that sort of thing.
```
          754678 function calls (748004 primitive calls) in 1.791 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.040    0.040    1.792    1.792 <string>:1(<module>)
        1    0.000    0.000    1.752    1.752 fixtures.py:76(__call__)
        1    0.000    0.000    1.040    1.040 fixtures.py:120(get_xml_nodes)
        1    0.048    0.048    1.038    1.038 fixtures.py:131(_get_fixture_node)
     2769    0.066    0.000    0.899    0.000 fixtures.py:330(_fill_in_location_element)
        1    0.001    0.001    0.675    0.675 fixtures.py:206(get_all_locations_to_sync)
    22/16    0.003    0.000    0.648    0.041 query.py:1085(_fetch_all)
     11/8    0.000    0.000    0.629    0.079 query.py:241(__iter__)
    55405    0.421    0.000    0.544    0.000 ElementTree.py:207(__init__)
     2817    0.008    0.000    0.504    0.000 query.py:48(__iter__)
     2769    0.121    0.000    0.387    0.000 fixtures.py:313(_get_metadata_node)
     2803    0.012    0.000    0.308    0.000 base.py:559(from_db)
     2775    0.010    0.000    0.292    0.000 models.py:387(__init__)
        1    0.001    0.001    0.244    0.244 fixtures.py:265(_get_include_without_expanding_locations)
     2803    0.075    0.000    0.236    0.000 base.py:457(__init__)
    61527    0.054    0.000    0.152    0.000 {setattr}
        3    0.000    0.000    0.139    0.046 query.py:663(_prefetch_related_objects)
        3    0.004    0.001    0.139    0.046 query.py:1387(prefetch_related_objects)
        3    0.007    0.002    0.133    0.044 query.py:1557(prefetch_one_level)
    55403    0.075    0.000    0.113    0.000 ElementTree.py:300(append)
       20    0.000    0.000    0.108    0.005 compiler.py:808(execute_sql)
     8320    0.019    0.000    0.104    0.000 related.py:597(get_local_related_value)
     2775    0.006    0.000    0.097    0.000 subclassing.py:36(__set__)
     3657    0.004    0.000    0.095    0.000 compiler.py:783(results_iter)
     2775    0.012    0.000    0.091    0.000 fields.py:64(pre_init)
       20    0.000    0.000    0.085    0.004 utils.py:58(execute)
       20    0.084    0.004    0.084    0.004 {method 'execute' of 'psycopg2.extensions.cursor' objects}
    55673    0.082    0.000    0.082    0.000 {method 'copy' of 'dict' objects}
       72    0.001    0.000    0.080    0.001 compiler.py:1231(cursor_iter)
       72    0.000    0.000    0.080    0.001 compiler.py:1237(<lambda>)
        3    0.002    0.001    0.079    0.026 related_descriptors.py:122(get_prefetch_queryset)
       73    0.000    0.000    0.079    0.001 utils.py:99(inner)
       72    0.028    0.000    0.078    0.001 {method 'fetchmany' of 'psycopg2.extensions.cursor' objects}
     2775    0.005    0.000    0.066    0.000 __init__.py:293(loads)
     2775    0.015    0.000    0.060    0.000 decoder.py:361(decode)
61072/55528    0.051    0.000    0.059    0.000 {getattr}
    18318    0.027    0.000    0.056    0.000 related_descriptors.py:153(__get__)
     8320    0.027    0.000    0.055    0.000 related.py:589(local_related_fields)
```

### CODE REVIEW

File: corehq/apps/locations/fixtures.py [hunk 1]
```diff
@@ -182,34 +182,44 @@ def should_sync_flat_fixture(project):
...
-
-        user_locations = set(user.get_sql_locations(user.domain))
-        for user_location in user_locations:
-            location_type = user_location.location_type
-            expand_from = location_type.expand_from or location_type
-            expand_to_level = (SQLLocation.active_objects
-                               .filter(domain__exact=user.domain, location_type=location_type.expand_to)
-                               .values_list('level', flat=True)
-                               .first())
+    return LocationSet(get_location_fixture_queryset(user))
```
Comment (esoergel) [inline corehq/apps/locations/fixtures.py]: I recommend reading this commit (Add in notes and general structure) with [`?w=1`](https://github.com/dimagi/commcare-hq/pull/16806/commits/354e0fc922d39f3412155e4f86ffec0db398ce7b?w=1)

File: corehq/apps/locations/fixtures.py [hunk 2]
```diff
@@ -250,20 +246,15 @@ def _get_include_without_expanding_locations(domain, location_type):
...
     if include_without_expanding is not None:
-        forced_location_level = set(
-            SQLLocation.active_objects.
-            filter(domain__exact=domain, location_type=include_without_expanding).
-            values_list('level', flat=True)
-        ) or None
+        forced_location_level = (SQLLocation.active_objects
+                                 .filter(domain__exact=domain, location_type=include_without_expanding)
+                                 .values_list('level', flat=True)
+                                 .first())
```
Comment (esoergel) [inline corehq/apps/locations/fixtures.py]: Similar to an earlier change - just use the first matching location to get the level, no need to look at all of them.

File: corehq/apps/locations/fixtures.py [hunk 3]
```diff
@@ -241,20 +240,28 @@ def _get_children(domain, root, expand_to_level):
...
+    forced_levels = (SQLLocation.active_objects
+                     .filter(domain__exact=domain,
+                             location_type__in=location_types)
+                     .values_list('level', flat=True)
+                     .order_by('level')
+                     .distinct('level'))
+    if forced_levels:
+        return (SQLLocation.active_objects
+                .filter(domain__exact=domain,
+                        level__lte=max(forced_levels)))
```
Comment (esoergel) [inline corehq/apps/locations/fixtures.py]: this only needs to be called once, with the max level, not once for each forced level
```
locations_lte3 | locations_lte2 == locations_lte3
```

File: corehq/apps/locations/fixtures.py [hunk 4]
```diff
@@ -241,20 +240,28 @@ def _get_children(domain, root, expand_to_level):
...
-            return (SQLLocation.active_objects
-                    .filter(domain__exact=domain,
-                            level__lte=forced_location_level))
-    return SQLLocation.objects.none()
+    # all loctypes to include, based on all assigned location types
+    location_types = (LocationType.objects
+                      .filter(sqllocation__in=assigned_locations)
+                      .exclude(include_without_expanding=None)
+                      .distinct()
+                      .values_list('include_without_expanding', flat=True))
```
Comment (proteusvacuum) [inline corehq/apps/locations/fixtures.py]:
```py
[
assigned_location.location_type.include_without_expanding_id for assigned_location in assigned_locations 
if assigned_location.location_type.include_without_expanding_id is not None
]
```

**Review by proteusvacuum — APPROVED** [review]

**Pull request merged**